File: Tellma/ClientApp/src/app/data/entities/reconciliation.ts
import { EntityForSave } from './base/entity-for-save';
import { EntryForReconciliation } from './entry-for-reconciliation';
import { ExternalEntry } from './external-entry';

// tslint:disable:max-line-length
export interface ReconciliationForSave<TEntry = ReconciliationEntryForSave, TExternalEntry = ReconciliationExternalEntryForSave> extends EntityForSave {
    Entries?: TEntry[];
    ExternalEntries?: TExternalEntry[];
}

export interface Reconciliation extends ReconciliationForSave<ReconciliationEntry, ReconciliationExternalEntry> {
    CreatedAt?: string;
    CreatedById?: number;
}

export interface ReconciliationEntryForSave extends EntityForSave {
    EntryId?: number;
}

export interface ReconciliationEntry extends ReconciliationEntryForSave {
    Entry?: EntryForReconciliation;
    CreatedAt?: string;
    CreatedById?: number;
}

export interface ReconciliationExternalEntryForSave extends EntityForSave {
    ExternalEntryId?: number;
    ExternalEntryIndex?: number;
}

export interface ReconciliationExternalEntry extends ReconciliationExternalEntryForSave {
    ExternalEntry?: ExternalEntry;
    CreatedAt?: string;
    CreatedById?: number;
}

File: Tellma/Entities/Reconciliation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace Tellma.Entities
{
    public class ReconciliationForSave<TEntry, TExternalEntry> : EntityWithKey<int>
    {
        public List<TEntry> Entries { get; set; }
        public List<TExternalEntry> ExternalEntries { get; set; }
    }

    public class ReconciliationForSave : ReconciliationForSave<ReconciliationEntryForSave, ReconciliationExternalEntryForSave>
    {
    }

    public class Reconciliation : ReconciliationForSave<ReconciliationEntry, ReconciliationExternalEntry>
    {

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }
    }

    public class ReconciliationEntryForSave : EntityWithKey<int>
    {
        [Required]
        public int? EntryId { get; set; }
    }

    public class ReconciliationEntry : ReconciliationEntryForSave
    {
        [ForeignKey(nameof(EntryId))]
        public Entry Entry { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }
    }

    public class ReconciliationExternalEntryForSave : EntityWithKey<int>
    {
        public int? ExternalEntryId { get; set; }

        [NotMapped]
        public int? ExternalEntryIndex { get; set; }
    }

    public class ReconciliationExternalEntry : ReconciliationExternalEntryForSave
    {
        [ForeignKey(nameof(ExternalEntryId))]
        public ExternalEntry ExternalEntry { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "CreatedBy")]
        [ForeignKey(nameof(CreatedById))]
        public User CreatedBy { get; set; }
    }
}
